### PR TITLE
Summary based on HdrHistogram

### PR DESCRIPTION
**Pull request by lavrov**
This work supersedes #484. The main difference is that HdrHistogram implementation is copy-pasted as is to remove external dependency.

### CODE REVIEW

File: simpleclient/src/test/java/io/prometheus/client/SummaryTest.java [hunk 1]
```diff
@@ -134,55 +219,69 @@ public Integer call() {
...
   @Test
-  public void noLabelsDefaultZeroValue() {
+  public void testNoLabels() {
     assertEquals(0.0, getCount(), .001);
     assertEquals(0.0, getSum(), .001);
-  }
+    assertEquals(null, getMin());
+    assertEquals(null, getMax());
 
-  private Double getLabelsCount(String labelValue) {
```
Comment (brian-brazil) [inline simpleclient/src/test/java/io/prometheus/client/SummaryTest.java:149]: Please don't change this. It makes the diff larger than it needs to be, and is also harder to understand.

File: simpleclient/src/main/java/io/prometheus/client/TimeWindowQuantiles.java [hunk 2]
```diff
@@ -1,54 +1,99 @@
...
 class TimeWindowQuantiles {
 
-  private final Quantile[] quantiles;
-  private final CKMSQuantiles[] ringBuffer;
-  private int currentBucket;
-  private long lastRotateTimestampMillis;
-  private final long durationBetweenRotatesMillis;
+  private final long highestToLowestValueRatio;
+  private final int numberOfSignificantValueDigits;
+  private final AtomicReference<ConcurrentDoubleHistogram[]> buckets;
```
Comment (brian-brazil) [inline simpleclient/src/main/java/io/prometheus/client/TimeWindowQuantiles.java:18]: Don't interleave config and state variables, it makes it harder to understand.

File: simpleclient/src/main/java/io/prometheus/client/TimeWindowQuantiles.java [hunk 3]
```diff
@@ -1,54 +1,99 @@
...
+
+    return buckets.get()[0]; // oldest bucket
+  }
+
+  private void rotate() {
+    // On concurrent `rotate` and `rotate`:
+    //  - `currentTime` is cached to reduce thread contention.
+    //  - `lastRotate` is used to ensure the correct number of rotations.
+
+    // Correctness is guaranteed by `volatile` memory access ordering and visibility semantics.
```
Comment (brian-brazil) [inline simpleclient/src/main/java/io/prometheus/client/TimeWindowQuantiles.java:77]: I don't see any use of volatile here. Are you sure this is right?

**Review by brian-brazil — COMMENTED** [review]
There's several unrelated changes in here, please keep the diff minimal with such a large change.

File: pom.xml [hunk 4]
```diff
@@ -136,8 +136,8 @@
                 <artifactId>maven-compiler-plugin</artifactId>
                 <version>3.8.0</version>
                 <configuration>
-                    <source>1.6</source>
-                    <target>1.6</target>
+                    <source>1.8</source>
```
[the remark this answers is not in the document]
Comment (lavrov) [reply pom.xml:139]: The change was necessary to compile HdrHistogram code. I will remove usages of 1.8 features.

File: simpleclient_common/src/test/java/io/prometheus/client/exporter/common/TextFormatTest.java [hunk 5]
```diff
@@ -104,6 +104,8 @@ public void testSummaryOutputWithQuantiles() throws IOException {
             + "labelsAndQuantiles{l=\"a\",quantile=\"0.5\",} 2.0\n"
             + "labelsAndQuantiles{l=\"a\",quantile=\"0.9\",} 2.0\n"
             + "labelsAndQuantiles{l=\"a\",quantile=\"0.99\",} 2.0\n"
+            + "labelsAndQuantiles_min{l=\"a\",} 2.0\n"
```
[the remark this answers is not in the document]
Comment (lavrov) [reply simpleclient_common/src/test/java/io/prometheus/client/exporter/common/TextFormatTest.java:107]: What it should look like?

[the remark this answers is not in the document]
Comment (brian-brazil) [reply simpleclient_common/src/test/java/io/prometheus/client/exporter/common/TextFormatTest.java:107]: The same as it does now.

File: simpleclient/src/main/java/io/prometheus/client/Summary.java [hunk 6]
```diff
@@ -74,52 +76,83 @@
...
-      if (error < 0.0 || error > 1.0) {
-        throw new IllegalArgumentException("Error " + error + " invalid: Expected number between 0.0 and 1.0.");
+      quantiles.add(quantile);
+      return this;
+    }
+
+    // backwards compatibility
+    public Builder quantile(double quantile, double error) {
+      this.quantile(quantile);
+      this.numberOfSignificantValueDigits(Math.max(this.numberOfSignificantValueDigits, (int)-Math.log10(error)));
```
[the remark this answers is not in the document]
Comment (lavrov) [reply simpleclient/src/main/java/io/prometheus/client/Summary.java:125]: Maximum value will be chosen.

File: simpleclient/src/main/java/io/prometheus/client/TimeWindowQuantiles.java (hunk 3 above)
Comment (lavrov) [reply simpleclient/src/main/java/io/prometheus/client/TimeWindowQuantiles.java]: Volatile inside atomic values was meant. I changed it to `atomic`.

File: simpleclient_common/src/test/java/io/prometheus/client/exporter/common/TextFormatTest.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (lavrov) [reply simpleclient_common/src/test/java/io/prometheus/client/exporter/common/TextFormatTest.java:107]: Do you mean that `min` and `max` should be calculated at all?

[the remark this answers is not in the document]
Comment (brian-brazil) [reply simpleclient_common/src/test/java/io/prometheus/client/exporter/common/TextFormatTest.java:107]: Yes, they should not be exposed.

[the remark this answers is not in the document]
Comment (lavrov) [reply simpleclient_common/src/test/java/io/prometheus/client/exporter/common/TextFormatTest.java]: Done, removed.

**Pull request closed without merging**